### PR TITLE
TC: support `...` patterns in constructor and tuple patterns

### LINKED ISSUE #440: Deal with `...` within tuple patterns

### LINKED ISSUE #441: Support `...` within constructor patterns

### PR DESCRIPTION
**Pull request by feds01**
This PR adds support for `...` patterns within tuple and constructor patterns. This is implemented by erasing `...` the patterns from these patterns and filling in any fields that they capture with a wildcard.

Additionally, if the user specified a bind on the `...` pattern, then the captured members are placed in a tuple that can be accessed in the proceeding scope. For example:

```rs
Foo := struct(
    age: i32,
    sex: char,
    name: str,
);

x := Foo::default();
Foo(age, ...other) := x; // other: (name: str, sex: char)
```

`Foo` is filled in as `Foo(age = _, sex = _, name = str )`. A similar process occurs for tuples and enums.

To be done:
- Figure out concrete rules for named/un-named argument placement within patterns/functions/tuples/etc.
- Code organisation: where should all the transformation actually occur (including field resolution)?
- Implement this for enum variants once enums are represented as unions.
- Make `ConstructorPat` subject always `ConstructedTerm` or `EnumVaraint`.

closes #440 
closes #441

### CODE REVIEW

File: compiler/hash-typecheck/src/ops/pats.rs [hunk 1]
```diff
@@ -133,6 +132,568 @@ impl<'tc> PatMatcher<'tc> {
...
+    ///
+    /// ```ignore
+    /// Foo := enum(
+    ///     Bar(i32, str, char),
+    /// );
+    ///
+    /// x := Foo::default();
+    ///
+    /// Foo::Bar(num, ...other) := x; // other: (str, char)
+    /// ```
```
Comment (kontheocharis) [inline compiler/hash-typecheck/src/ops/pats.rs:230]: This isn't specific to enums---also applies to nameless struct fields

File: compiler/hash-typecheck/src/ops/pats.rs [hunk 2]
```diff
@@ -133,6 +132,568 @@ impl<'tc> PatMatcher<'tc> {
...
+            return Ok(None);
+        }
+        let (_pos, name, spread_id) = spread_pat.unwrap();
+
+        let subject_members = match self.oracle().term_as_nominal_def(subject).unwrap() {
+            NominalDef::Struct(struct_def) => match struct_def.fields {
+                StructFields::Explicit(fields) => fields,
+                _ => unreachable!(),
+            },
+            NominalDef::Enum(_) => unreachable!(),
```
Comment (kontheocharis) [inline compiler/hash-typecheck/src/ops/pats.rs:257]: Maybe leave a comment here?

Comment (kontheocharis) [reply compiler/hash-typecheck/src/ops/pats.rs:257]: Can use .term_as_struct_def

File: compiler/hash-typecheck/src/ops/validate.rs [hunk 3]
```diff
@@ -1238,4 +1238,183 @@ impl<'tc> Validator<'tc> {
...
+    ///
+    /// If the constructor subject is a `struct`, the argument names are
+    /// `inferred`  from the pattern bindings, if they have no names, then
+    /// they are assumed to be the name of the position that they are
+    /// specified at. The function will automatically apply the conversion
+    /// and amend the stored arguments with the inferred names.
+    ///
+    /// If the constructor subject is an `enum`, then standard argument position
+    /// rules are applied.
+    pub(crate) fn validate_constructor_pat(&self, pat: PatId) -> TcResult<()> {
```
Comment (kontheocharis) [inline compiler/hash-typecheck/src/ops/validate.rs:1297]: Add @@Todo to write equivalent methods for expression literals

File: compiler/hash-typecheck/src/ops/validate.rs [hunk 4]
```diff
@@ -1238,4 +1238,183 @@ impl<'tc> Validator<'tc> {
...
+                            // the index from the end of the parameter list as if it
+                            // was in a tuple.
+                            let offset = if has_spread_pat {
+                                members.len() - (pat_args_length - index)
+                            } else {
+                                index
+                            };
+
+                            // Guaranteed to have a name since this is a struct definition.
+                            members.positional()[offset].name.unwrap()
```
Comment (kontheocharis) [inline compiler/hash-typecheck/src/ops/validate.rs:1378]: This generally shouldn't be assumed

File: compiler/hash-typecheck/src/storage/param_list.rs [hunk 5]
```diff
@@ -50,6 +50,11 @@ impl<ParamListKey: SequenceStoreKey, ParamType: GetNameOpt + Clone>
         self.origins.get(key).unwrap_or(ParamOrigin::Unknown)
     }
 
+    /// Get the size of the parameter list
+    pub fn get_size(&self, key: ParamListKey) -> usize {
+        self.map_as_param_list_fast(key, |list| list.len())
+    }
+
```
Comment (kontheocharis) [inline compiler/hash-typecheck/src/storage/param_list.rs:57]: Remove

**Review by kontheocharis — CHANGES_REQUESTED** [review]

File: compiler/hash-typecheck/src/ops/params.rs [hunk 6]
```diff
@@ -197,3 +197,57 @@ pub(crate) fn validate_param_list_ordering<T: Clone + GetNameOpt>(
...
+                    return Err(TcError::ParamGivenTwice { param_kind: origin, index });
+                } else {
+                    params_used.insert(found_index);
+                }
+            }
+        }
+    }
+
+    Ok(())
+}
```
[the remark this answers is not in the document]
Comment (feds01) [reply compiler/hash-typecheck/src/ops/params.rs:228]: This doesn't actually modify the structure...

[the remark this answers is not in the document]
Comment (kontheocharis) [reply compiler/hash-typecheck/src/ops/params.rs:233]: Yes sorry, I was referring to the point in the code where this actually happens.

**Review by kontheocharis — APPROVED** [review]

**Pull request merged**